File: python/marker_analysis.py
import itertools
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import scanpy as sc
import episcanpy as epi



def top_markers(adata, ntop=100):
    """
    Find the top biomarkers in a list of differentially distributed
    features in an AnnData object from scanpy.tl.rank_genes_groups().
    - adata: AnnData object.
    - ntop: Number of top markers to keep
    """

    markers = None

    # If processed with Scanpy
    if 'rank_genes_groups' in adata.uns:
        markers = pd.DataFrame(
            adata.uns['rank_genes_groups']['names']
        ).head(ntop)

    # Else if processed with Episcanpy
    elif 'rank_features_groups' in adata.uns:
        markers = pd.DataFrame(
            adata.uns['rank_features_groups']['names']
        ).head(ntop)
    else:
        print('No rank_genes_groups variable found in the submitted ',
              'anndata. Please run scanpy.tl.rank_genes_groups() to ',
              'find group markers.')

    return(markers)



def plot_jaccardindex(data, xlab='x', ylab='y'):
    """
    Plot the jaccard index matrix as a heatmap.
    """

    data = data.round(decimals=2)
    # This will use the same order for the labels
    data = data.sort_index(0).sort_index(1)
    # This will reverse the matrix so the diagonal looks good
    data = data[::-1]

    plot = sns.heatmap(data, annot = True, vmin = 0.,
                       vmax = 1., cmap = 'Reds')
    plt.xticks(rotation = 70)
    plt.xlabel(xlab, fontsize = 30)
    plt.ylabel(ylab, fontsize = 30)
    plt.title('Jaccard Index scores', fontsize = 40, pad = 20)
    plt.show()

    return(plot)



def matchscore(ref_markers, obs_markers, plot=True,
              xlab='My dataset', ylab='Reference dataset'):
    """
    Analysis of similarity between a set of markers found in
    a reference dataset and a set of markers from the observed
    query dataset.
    - ref_markers: List of markers from the reference
    - obs_markers: List of markers from the observed query
    - plot: Whether or not to plot the jaccard index matrix
    """

    score = 0
    mat = []

    for ref_cluster in ref_markers:
        ref_len = len(ref_markers[ref_cluster])

        jacc_index = []

        for obs_cluster in obs_markers:

            obs_len = len(obs_markers[obs_cluster])
            shared = len(set(ref_markers[ref_cluster]).intersection(
                obs_markers[obs_cluster]))
            jacc = shared / (ref_len + obs_len - shared)
            jacc_index.append(jacc)

        score += np.max(jacc_index)
        mat.append(jacc_index)

    mat = pd.DataFrame(mat, index = list(ref_markers.columns),
                       columns = list(obs_markers.columns))

    score = score / ref_markers.shape[1]
    max_ji = mat.max()
    anno_lab = mat.idxmax()
    if plot:
        plot_jaccardindex(mat, xlab=xlab, ylab=ylab)

    return((score, anno_lab, max_ji, mat))



def find_common_genes(ref, sample, target_n_genes=2000):
    """
    Find a common set of features between two scRNA datasets
    using the highly variable genes.
    - ref: Reference AnnData object
    - sample: Query AnnData object
    - target_n_genes: Target number of common genes to find
    """

    print(f'Dimensions before filtering: {ref.shape} and {sample.shape}\n')

    n = 5000
    common = []
    while len(common) < target_n_genes:

        print(f'Looking for {n} HVG')
        sc.pp.highly_variable_genes(ref, n_top_genes=n)
        sc.pp.highly_variable_genes(sample, n_top_genes=n)

        ref_genes = ref.var.highly_variable[ref.var.highly_variable == True].index
        sample_genes = sample.var.highly_variable[sample.var.highly_variable == True].index

        common = list(set(sample_genes).intersection(ref_genes))
        print(f'Found {len(common)} genes in common')
        n += 1000

    print('Filtering the data to these features and scaling')

    ref = ref[:, common]
    sample = sample[:, common]
    sc.pp.scale(ref, max_value=10)
    sc.pp.scale(sample, max_value=10)
    print(f'\nDimensions after filtering: {ref.shape} and {sample.shape}')

    return ref, sample



def find_common_variable_peaks(ref, sample, target_n_peaks=5000):
    """
    Find a common set of features between two scATAC datasets
    using the highly variable peaks.
    - ref: Reference AnnData object
    - sample: Query AnnData object
    - target_n_genes: Target number of common peaks to find
    """

    print('Dimensions before filtering: ',
          f'{ref.shape} and {sample.shape}\n')

    n = target_n_peaks
    common = []
    while len(common) < target_n_peaks:

        print(f'Looking for {n} HVP')
        ref_var = epi.pp.select_var_feature(
            ref, nb_features=n, show=False, copy=True)
        sample_var = epi.pp.select_var_feature(
            sample, nb_features=n, show=False, copy=True)

        ref_genes = list(ref_var.var.index)
        sample_genes = list(sample_var.var.index)

        common = list(set(sample_genes).intersection(ref_genes))
        print(f'Found {len(common)} peaks in common')
        n += 1000

    print('Filtering the data to these features and scaling')

    ref = ref[:, common]
    sample = sample[:, common]
    sc.pp.scale(ref, max_value=10)
    sc.pp.scale(sample, max_value=10)
    print('Dimensions after filtering: ',
          f'{ref.shape} and {sample.shape}\n')

    return ref, sample



def find_refmarkers_in_variable_peaks(ref, sample, target_n_peaks=5000):
    """
    Find a common set of features between two scATAC datasets
    using the marker peaks from the reference and the highly
    variable peaks from the query sample.
    - ref: Reference AnnData object
    - sample: Query AnnData object
    - target_n_genes: Target number of common peaks to find
    """

    print('Dimensions before filtering: ',
          f'{ref.shape} and {sample.shape}\n')

    dfn = ref.obs['predicted.id'].value_counts() > 20
    keep_celltypes = dfn[dfn == True].index.to_list()
    ref = ref[ref.obs['predicted.id'].isin(keep_celltypes)]
    sc.pp.normalize_total(ref)
    epi.pp.log1p(ref)

    sc.pp.normalize_total(sample)
    epi.pp.log1p(sample)

    n_cat = len(ref.obs['predicted.id'].cat.categories)
    n = int(round(target_n_peaks*2/n_cat))
    common = []

    while len(common) < target_n_peaks:

        print(f'Looking for {n} marker features from each label')
        epi.tl.rank_features(ref, 'predicted.id', omic="ATAC",
                             use_raw=False, n_features=n)
        marker_peaks = list(ref.uns['rank_features_groups']['names'])
        marker_peaks = list(itertools.chain.from_iterable(marker_peaks))
        print(len(marker_peaks))

        sample_peaks = epi.pp.select_var_feature(
            sample, nb_features=len(marker_peaks), show=False, copy=True)
        sample_peaks = list(sample_peaks.var.index)

        common = list(set(marker_peaks).intersection(sample_peaks))
        print(f'Found {len(common)} peaks in common')
        n += 100

    print('Filtering the data to these features and scaling')

    ref = ref[:, common]
    sample = sample[:, common]
    sc.pp.scale(ref, max_value=10)
    sc.pp.scale(sample, max_value=10)
    print('Dimensions after filtering: ',
          f'{ref.shape} and {sample.shape}\n')

    return ref, sample

File: R/marker_analysis.R


top_markers <- function(markers, ntop=10) {

    c_names <- levels(markers$cluster)

    top <- lapply(c_names, function(x) markers$gene[markers$cluster == x][1:ntop])
    top <- lapply(top, function(x) x[!is.na(x)])
    names(top) <- c_names
    return(top)
}
                  

find_common_variable_genes <- function(ref, sample, ref.assay="RNA",
                                       sample.assay="RNA", target_n_genes=4000) {

    DefaultAssay(ref) <- ref.assay
    DefaultAssay(sample) <- sample.assay

    n <- 10000
    common <- c()
    while (length(common) < target_n_genes) {
        
        print(glue('Looking for {n} HVG'))
        ref <- FindVariableFeatures(ref, selection.method = "vst", nfeatures = n)
        sample <- FindVariableFeatures(sample, selection.method = "vst", nfeatures = n)
        
        ref_genes <- VariableFeatures(ref)
        sample_genes <- VariableFeatures(sample)
        
        common <- intersect(ref_genes, sample_genes)
        print(glue('Found {length(common)} genes in common'))
        n <- n + 1000
    }

    return(common)
}
                  

find_refmarkers_in_variable_genes <- function(ref.markers, sample, sample.assay="RNA", target_n_genes=2000) {

    DefaultAssay(sample) <- sample.assay
    n_cat <- length(levels(ref.markers$cluster))
    n <- as.integer(round(target_n_genes*2/n_cat))

    common <- c()
    
    prev.n.variable <- 0
    max.reached <- FALSE
    
    while ((length(common) < target_n_genes) & (!max.reached)) {
        
        print(glue('Looking for {n} marker genes from each label'))
        ref_genes <- top_markers(ref.markers, ntop = n)
        ref_genes <- unname(unlist(ref_genes))

        sample <- FindVariableFeatures(sample, selection.method = "vst", nfeatures = length(ref_genes))

        common <- intersect(ref_genes, VariableFeatures(sample))
        print(glue('Found {length(common)} genes in common'))
        n <- n + 100

        if (length(VariableFeatures(sample)) == prev.n.variable) {
            print('No more DEG found, returning last result')
            max.reached <- TRUE
        }
        
        prev.n.variable <- length(VariableFeatures(sample))
    }

    return(common)
}
                  
                  
find_refmarkers_in_genes <- function(ref.markers, sample, sample.assay="RNA", target_n_genes=2000) {

    DefaultAssay(sample) <- sample.assay
    n_cat <- length(levels(ref.markers$cluster))
    n <- as.integer(round(target_n_genes/n_cat))

    common <- c()
    max.reached <- FALSE
    
    while ((length(common) < target_n_genes) & (!max.reached)) {
        
        print(glue('Looking for {n} marker genes from each label'))
        ref_genes <- top_markers(ref.markers, ntop = n)
        ref_genes <- unname(unlist(ref_genes))

        common <- intersect(ref_genes, rownames(sample))
        print(glue('Found {length(common)} genes in common'))
        n <- n + 100
        
        if (length(ref.markers$gene) == length(ref_genes)) {
            print('No more markers to go through, returning last result')
            max.reached <- TRUE
        }
    }

    return(common)
}
                  
                  
                  
                  
find_common_top_peaks <- function(ref, sample, ref.assay="peaks",
                                  sample.assay="peaks", target_n_peaks=6000) {

    DefaultAssay(ref) <- ref.assay
    DefaultAssay(sample) <- sample.assay
    
    ref <- FindTopFeatures(ref)
    sample <- FindTopFeatures(sample)

    n <- target_n_peaks
    common <- c()
    while (length(common) < target_n_peaks) {
        
        print(glue('Looking for {n} top peaks'))
        
        ref_peaks <- ref@assays$peaks@meta.features %>% 
                        arrange(desc(count)) %>% 
                        top_n(n)
        ref_peaks <- rownames(ref_peaks)
        
        sample_peaks <- sample@assays$peaks@meta.features %>% 
                            arrange(desc(count)) %>% 
                            top_n(n)
        sample_peaks <- rownames(sample_peaks)
        
        common <- intersect(ref_peaks, sample_peaks)
        print(glue('Found {length(common)} top peaks in common'))
        n <- n + 1000
    }

    return(common)
}
     
                  
find_refmarkers_in_top_peaks <- function(ref.markers, sample, sample.assay="peaks", target_n_peaks=6000) {

    DefaultAssay(sample) <- sample.assay
    sample <- FindTopFeatures(sample)

    n <- target_n_peaks

    ref_peaks <- top_markers(ref.markers, ntop = length(ref.markers$gene))
    ref_peaks <- unname(unlist(ref_peaks))

    common <- c()
    prev.n.variable <- 0
    max.reached <- FALSE
    
    while ((length(common) < target_n_peaks) & (!max.reached)) {
        
        print(glue('Looking for {n} marker peaks from each label'))

        sample_peaks <- sample@assays$peaks@meta.features %>% 
                            arrange(desc(count)) %>% 
                            top_n(n)
        sample_peaks <- rownames(sample_peaks)

        common <- intersect(ref_peaks, sample_peaks)
        print(glue('Found {length(common)} peaks in common'))
        n <- n + 3000

        if (length(sample_peaks) == prev.n.variable) {
            print('No more DAR found, returning last result')
            max.reached <- TRUE
        }
        
        prev.n.variable <- length(sample_peaks)
    }

    return(common)
}

                  
                  
find_refmarkers_in_peaks <- function(ref.markers, sample, sample.assay="peaks", target_n_peaks=6000) {

    DefaultAssay(sample) <- sample.assay
    n_cat <- length(levels(ref.markers$cluster))
    n <- as.integer(round(target_n_peaks/n_cat))

    common <- c()
    max.reached <- FALSE
    
    while ((length(common) < target_n_peaks) & (!max.reached)) {
        
        print(glue('Looking for {n} marker peaks from each label'))
        ref_genes <- top_markers(ref.markers, ntop = n)
        ref_genes <- unname(unlist(ref_genes))

        common <- intersect(ref_genes, rownames(sample))
        print(glue('Found {length(common)} peaks in common'))
        n <- n + 100
        
        if (length(ref.markers$gene) == length(ref_genes)) {
            print('No more markers to go through, returning last result')
            max.reached <- TRUE
        }
    }

    return(common)
}
                  
                  
                  
summary_ggplot <- function(data, ylab, xlab) {
    
    df <- data.frame(t(data), check.names = FALSE, check.rows = FALSE)
    df <- melt(cbind(pred_id = rownames(df), df), id = "pred_id")
    df$pred_id <- factor(df$pred_id, levels = levels(df$variable))

    gg <- ggplot(df, aes_string(x = "pred_id", y = "variable", fill = "value")) + 
        labs(x = xlab, y = ylab) + geom_tile(aes_string(fill = "value")) +
        geom_text(aes(label = round(value, 2))) + scale_fill_viridis_c() +
        theme(
            axis.text.x = element_text(angle = 30, hjust = 1, size = 16),
            axis.text.y = element_text(size = 16),
            axis.title = element_text(size = 16)
        )
    
    return(gg)
}

                  
                  
matchSCore2 <- function(ref.markers, query.markers,
                        xlab="Predicted labels", ylab="Reference labels") {
    
    score <- 0
    lab.ref <- seq(1:length(ref.markers))
    lab.obs <- seq(1:length(query.markers))

    JI_matrix <- vector()
    
    for (ref.cluster in ref.markers) {
        
        JI <- vector()
        
        for (q.cluster in query.markers) {
        
            I <- length(intersect(ref.cluster, q.cluster))
            J <- I / (length(ref.cluster) + length(q.cluster) - I)
            JI <- append(JI, J, length(JI))
        }
        
        score <- sum(score, max(JI))
        JI_matrix <- rbind(JI_matrix, JI)
    }
    
    score <- score / length(ref.markers)
    colnames(JI_matrix) <- names(query.markers)
    rownames(JI_matrix) <- names(ref.markers)

    gg <- summary_ggplot(data = JI_matrix, ylab, xlab)

    return(
        list(
            matchScore = score,
            labels = apply(JI_matrix, 2, which.max),
            max_JI = apply(JI_matrix, 2, max),
            JI.mat = JI_matrix,
            ggplot = gg
        )
    )
}
                  
                  
                  
